luaintro/springconfig.lua

```diff
@@ -147,3 +147,18 @@ if Spring.GetConfigInt("version", 0) < version then
 		Spring.SetConfigInt("UnitIconFadeVanish", 3000)
 	end
 end
+
+-- Configure sane keychain settings, this is to provide a standard experience
+-- for users that is acceptable
+local springKeyChainTimeout = 750 -- in ms
+local userKeyChainTimeout = Spring.GetConfigInt("KeyChainTimeout", springKeyChainTimeout)
+-- Only apply BARs default if current setting is equal to engine default
+-- Reason is engine is unable to distinguish between:
+--   - user configuring the setting to be equal to default
+--   - the actual setting being empty and engine using default
```

Comment:
could always do
```lua
local userKeyChainTimeout = Spring.GetConfigInt("KeyChainTimeout", springKeyChainTimeout)
if userKeyChainTimeout == springKeyChainTimeout then
    local nil_check = Spring.GetConfigInt("KeyChainTimeout", 1337)
    if nil_check == 1337 then
      -- it's picking the default each time so was nil/unset
    elseif nil_check == springKeyChainTimeout then
      -- it's explicitly set to the default
    else
      error("lol wtf")
    end
end

Reply:
possibly wrapped into something more general:
```lua
function Spring.Utilities.GetConfigIntNullable(key) -- returns value if set, else nil
  local value = Spring.GetConfigInt(key, 123)
  if value ~= 123 then
    return value
  end

  -- 
  local value2 = Spring.GetConfigInt(key, 456)
  if value2 == 456 then
    -- returns the default each time, has to be nil
    return nil
  elseif value2 == 123 then
    -- happens to be set to the first magic value
    return 123
  else
    error("Behaviour of `Spring.GetConfigInt` changed!")
  end
end

-- usage for our case
if not Spring.Utilities.GetConfigIntNullable("KeyChainTimeout") == nil then
  Spring.SetConfigInt("KeyChainTimeout", 333)
end

Reply:
Let me do a few more tests, maybe my assumptions were wrong

Reply:
Ok, they were not. You still need to know what's the engine default in game-specific code unfortunately.

Reply:
I simplified the code a bit though.

Reply:
Maybe the engine should do the above. See https://github.com/beyond-all-reason/spring/pull/556

Reply:
Ah, I see what you mean. Using two different 2 argument calls to check whether current is default or not as a workaround for current API.

That makes sense, although I'm not particularly worried at doing this for the moment. If I do it I'll send it as a separate PR which will have much increased scope.

Reply:
> You still need to know what's the engine default in game-specific code unfortunately.

Does that mean it returns (for the specific case of key timeout) 750 even if you pass a different default? Or does it return whatever you pass, but you still need to pass 750 just so the widget has the same behaviour as default engine behaviour?

Reply:
It resolves a category of issues related to boilerplate, but not the fundamental issue of engine still deleting your setting entries if they happen to be equal to default and not allowing games to know whether user just wants the default as a setting or user never configured it.

Reply:
> > You still need to know what's the engine default in game-specific code unfortunately.
> 
> Does that mean it returns (for the specific case of key timeout) 750 even if you pass a different default? Or does it return whatever you pass, but you still need to pass 750 just so the widget has the same behaviour as default engine behaviour?

KeyChainTimeout is unset on springsettings, running:

```lua
	local res = Spring.GetConfigInt('KeyChainTimeout')
	Spring.Echo(os.clock(), "Spring.GetConfigInt('KeyChainTimeout')", type(res), res)
	res = Spring.GetConfigInt('KeyChainTimeout', 1234)
	Spring.Echo(os.clock(), "Spring.GetConfigInt('KeyChainTimeout', 1234)", type(res), res)
```

Gives me: 

![2022-12-23T11:21:03,225894232-03:00](https://user-images.githubusercontent.com/347552/209351422-d9854a64-765a-4efb-bd13-e10e9971e670.png)

Reply:
So it indicates that you can in fact assess if the user has set the setting or not.

I find this behavior surprising in some senses, but not in others. Personally I find the whole thing broken.

It tells me your suggestion is in fact an improvement to workaround this brokenness that should be incorporated at some point.

Reply:
In any manner, I won't change the current PR to incorporate it. But keep it in mind as a further improvement.

To that end I created https://github.com/beyond-all-reason/Beyond-All-Reason/issues/1465, thanks for bringing this up.

Reply:
After some testing and the results above, it seems to me this might not be as achievable as I imagined (to know the engine default at runtime). I closed the issue. Engine should provide a saner API or perhaps a method to get the settings default.

Reply:
> So it indicates that you can in fact assess what is the default.

How do you know the 750 is the default and not just the user setting? If the default was actually 1234 but you set it to 750 manually then both of your calls (no second arg and 1234 as the second arg) would still look the same and both return 750.

Reply:
Exactly, as per my last message and closure of the aforementioned issue I think this is a problem with the engine API. Ideally it would allow users to configure settings to be equal to default _and_ expose an API that makes sense.

I'll resolve this conversation as unrelated to the scope of this PR unless you have anything to add/object to.

Reply:
> Maybe the engine should do the above. See [beyond-all-reason/spring#556](https://github.com/beyond-all-reason/spring/pull/556)

This looks good!